File: scripts/lib/atproto/diagnostics.py
"""Shared run context and diagnostics for ATProto e2e scripts.

The scenario runner and Python launchers use the same run-directory layout as
the Bash helpers in ``scripts/lib/common.sh``:

    /tmp/garazyk-atproto-e2e/<run-id>/
      logs/
      reports/
      diagnostics/

Diagnostics are intentionally file-based so shell, Python, Docker, and local
binary runs can all leave useful evidence without requiring a service to stay
alive after a failing test.
"""

from __future__ import annotations

import json
import os
import re
import shutil
import subprocess
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable, Mapping

import requests

from .config import SERVICE_URLS, find_project_root


_BASE_DIR = Path("/tmp/garazyk-atproto-e2e")
_SECRET_PATTERNS = [
    re.compile(r"(Authorization:\s*Bearer\s+)[A-Za-z0-9._~+/=-]+", re.IGNORECASE),
    re.compile(
        r'("(?:accessJwt|refreshJwt|token|password|secret|masterSecret|adminPassword)"\s*:\s*")[^"]+"',
        re.IGNORECASE,
    ),
    re.compile(r"((?:JWT|TOKEN|PASSWORD|SECRET|MASTER_SECRET|ADMIN_SECRET)=)\S+"),
]


@dataclass(frozen=True)
class E2ERunContext:
    """Resolved filesystem locations for one e2e run."""

    run_id: str
    run_dir: Path
    logs_dir: Path
    reports_dir: Path
    diagnostics_dir: Path
    pid_file: Path
    compose_project: str


def sanitize_run_id(value: str) -> str:
    """Return a compose/filesystem-safe run id."""
    cleaned = re.sub(r"[^A-Za-z0-9_.-]+", "-", value.strip()).strip("-").lower()
    return cleaned or default_run_id()


def default_run_id() -> str:
    """Return a unique default run id."""
    return f"{time.strftime('%Y%m%dT%H%M%SZ', time.gmtime())}-{os.getpid()}"


def create_run_context(
    run_id: str | None = None,
    diagnostics_dir: str | Path | None = None,
    run_dir: str | Path | None = None,
) -> E2ERunContext:
    """Create and export the shared run-directory context."""
    resolved_run_id = sanitize_run_id(
        run_id or os.environ.get("ATPROTO_E2E_RUN_ID", "") or default_run_id()
    )
    base_dir = Path(os.environ.get("ATPROTO_E2E_BASE_DIR", str(_BASE_DIR)))
    resolved_run_dir = Path(
        run_dir or os.environ.get("ATPROTO_E2E_RUN_DIR", base_dir / resolved_run_id)
    )
    resolved_logs_dir = Path(os.environ.get("ATPROTO_E2E_LOG_DIR", resolved_run_dir / "logs"))
    resolved_reports_dir = Path(
        os.environ.get("ATPROTO_E2E_REPORTS_DIR", resolved_run_dir / "reports")
    )
    resolved_diag_dir = Path(
        diagnostics_dir
        or os.environ.get("ATPROTO_E2E_DIAGNOSTICS_DIR", resolved_run_dir / "diagnostics")
    )
    resolved_pid_file = Path(
        os.environ.get("ATPROTO_E2E_PID_FILE", resolved_run_dir / "pids.txt")
    )
    compose_run_id = re.sub(r"[^a-z0-9-]+", "-", resolved_run_id.replace(".", "-").replace("_", "-"))
    compose_project = os.environ.get(
        "ATPROTO_E2E_COMPOSE_PROJECT", f"garazyk-e2e-{compose_run_id}"
    )

    for path in (resolved_run_dir, resolved_logs_dir, resolved_reports_dir, resolved_diag_dir):
        path.mkdir(parents=True, exist_ok=True)

    os.environ["ATPROTO_E2E_RUN_ID"] = resolved_run_id
    os.environ["ATPROTO_E2E_RUN_DIR"] = str(resolved_run_dir)
    os.environ["ATPROTO_E2E_LOG_DIR"] = str(resolved_logs_dir)
    os.environ["ATPROTO_E2E_REPORTS_DIR"] = str(resolved_reports_dir)
    os.environ["ATPROTO_E2E_DIAGNOSTICS_DIR"] = str(resolved_diag_dir)
    os.environ["ATPROTO_E2E_PID_FILE"] = str(resolved_pid_file)
    os.environ["ATPROTO_E2E_COMPOSE_PROJECT"] = compose_project

    return E2ERunContext(
        run_id=resolved_run_id,
        run_dir=resolved_run_dir,
        logs_dir=resolved_logs_dir,
        reports_dir=resolved_reports_dir,
        diagnostics_dir=resolved_diag_dir,
        pid_file=resolved_pid_file,
        compose_project=compose_project,
    )


def redact(value: str) -> str:
    """Redact common token/password shapes from diagnostic summaries."""
    redacted = value
    for pattern in _SECRET_PATTERNS:
        if "accessJwt" in pattern.pattern:
            redacted = pattern.sub(r'\1[REDACTED]"', redacted)
        else:
            redacted = pattern.sub(r"\1[REDACTED]", redacted)
    return redacted


def write_text(path: Path, text: str) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(redact(text), encoding="utf-8")


def collect_http_endpoint(
    output_dir: Path,
    name: str,
    url: str,
    headers: Mapping[str, str] | None = None,
    timeout: int = 8,
) -> None:
    """Capture one HTTP endpoint into a redacted diagnostic file."""
    target = output_dir / "http" / f"{name}.txt"
    try:
        resp = requests.get(url, headers=dict(headers or {}), timeout=timeout)
        body = resp.text[:50_000]
        text = (
            f"url={url}\n"
            f"http_status={resp.status_code}\n"
            f"content_type={resp.headers.get('Content-Type', '')}\n\n"
            f"{body}\n"
        )
    except requests.RequestException as exc:
        text = f"url={url}\nerror={exc}\n"
    write_text(target, text)


def collect_diagnostics(
    context: E2ERunContext,
    *,
    service_urls: Mapping[str, str] | None = None,
    appview_admin_secret: str | None = None,
    compose_dir: str | Path | None = None,
    compose_files: Iterable[str | Path] = (),
    compose_project: str | None = None,
    label: str = "atproto-e2e",
) -> Path:
    """Collect service health, logs, compose state, and run metadata."""
    output_dir = context.diagnostics_dir
    output_dir.mkdir(parents=True, exist_ok=True)
    urls = dict(service_urls or SERVICE_URLS)
    appview_secret = appview_admin_secret or os.environ.get(
        "APPVIEW_ADMIN_SECRET", "localdevadmin"
    )

    metadata = {
        "label": label,
        "run_id": context.run_id,
        "run_dir": str(context.run_dir),
        "diagnostics_dir": str(output_dir),
        "compose_project": compose_project or context.compose_project,
        "created_at_utc": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        "service_urls": urls,
    }
    try:
        root = find_project_root()
        metadata["repo_root"] = str(root)
        metadata["git_commit"] = subprocess.run(
            ["git", "-C", str(root), "rev-parse", "HEAD"],
            capture_output=True,
            text=True,
            check=False,
        ).stdout.strip()
        metadata["git_status"] = subprocess.run(
            ["git", "-C", str(root), "status", "--short"],
            capture_output=True,
            text=True,
            check=False,
        ).stdout.splitlines()
    except Exception as exc:  # pragma: no cover - diagnostic best effort
        metadata["git_error"] = str(exc)

    write_text(output_dir / "run-metadata.json", json.dumps(metadata, indent=2))

    if context.pid_file.exists():
        shutil.copy2(context.pid_file, output_dir / "pids.txt")
    if context.logs_dir.exists():
        logs_out = output_dir / "service-logs"
        logs_out.mkdir(exist_ok=True)
        for log_path in context.logs_dir.glob("*.log"):
            shutil.copy2(log_path, logs_out / log_path.name)

    collect_http_endpoint(output_dir, "plc-health", f"{urls.get('plc', '')}/_health")
    collect_http_endpoint(
        output_dir,
        "pds-describe-server",
        f"{urls.get('pds', '')}/xrpc/com.atproto.server.describeServer",
    )
    collect_http_endpoint(output_dir, "relay-health", f"{urls.get('relay', '')}/api/relay/health")
    collect_http_endpoint(
        output_dir, "relay-upstreams", f"{urls.get('relay', '')}/api/relay/upstreams"
    )
    collect_http_endpoint(
        output_dir,
        "appview-backfill-status",
        f"{urls.get('appview', '')}/admin/backfill/status",
        headers={"Authorization": f"Bearer {appview_secret}"},
    )
    collect_http_endpoint(
        output_dir,
        "pds2-describe-server",
        f"{urls.get('chat', '')}/xrpc/com.atproto.server.describeServer",
    )
    collect_http_endpoint(output_dir, "chat-health", f"{urls.get('chat', '')}/_health")
    collect_http_endpoint(output_dir, "video-health", f"{urls.get('video', '')}/_health")
    collect_http_endpoint(output_dir, "ui-admin", f"{urls.get('ui', '')}/admin")

    compose_paths = [Path(path) for path in compose_files]
    if compose_dir and compose_paths and shutil.which("docker"):
        docker_out = output_dir / "docker"
        docker_out.mkdir(exist_ok=True)
        cmd_base = ["docker", "compose"]
        project = compose_project or context.compose_project
        if project:
            cmd_base.extend(["-p", project])
        for path in compose_paths:
            cmd_base.extend(["-f", str(path)])
        for name, args in {
            "ps": ["ps", "--all"],
            "config": ["config"],
            "logs": ["logs", "--no-color", "--timestamps", "--tail=300"],
        }.items():
            result = subprocess.run(
                cmd_base + args,
                cwd=str(compose_dir),
                capture_output=True,
                text=True,
                check=False,
            )
            write_text(docker_out / f"{name}.txt", result.stdout + result.stderr)

    return output_dir

File: scripts/lib/deno/diagnostics.ts
import { join, resolve } from "@std/path";
import { copy, exists } from "@std/fs";
import { SERVICE_URLS } from "./config.ts";

const BASE_DIR = "/tmp/garazyk-atproto-e2e";

const SECRET_PATTERNS = [
  /(Authorization:\s*Bearer\s+)[A-Za-z0-9._~+/=-]+/gi,
  /("(?:accessJwt|refreshJwt|token|password|secret|masterSecret|adminPassword)"\s*:\s*")[^"]+"/gi,
  /((?:JWT|TOKEN|PASSWORD|SECRET|MASTER_SECRET|ADMIN_SECRET)=)\S+/g,
];

export interface E2ERunContext {
  runId: string;
  runDir: string;
  logsDir: string;
  reportsDir: string;
  diagnosticsDir: string;
  pidFile: string;
  composeProject: string;
}

function sanitizeRunId(value: string): string {
  const cleaned = value.trim().replace(/[^A-Za-z0-9_.-]+/g, "-").replace(/^-+|-+$/g, "")
    .toLowerCase();
  return cleaned || defaultRunId();
}

function defaultRunId(): string {
  const now = new Date().toISOString().replace(/[:.]/g, "").substring(0, 15) + "Z";
  return `${now}-${Deno.pid}`;
}

export async function createRunContext(
  runId?: string,
  diagnosticsDir?: string,
  runDir?: string,
): Promise<E2ERunContext> {
  const resolvedRunId = sanitizeRunId(
    runId || Deno.env.get("ATPROTO_E2E_RUN_ID") || defaultRunId(),
  );
  const baseDir = Deno.env.get("ATPROTO_E2E_BASE_DIR") || BASE_DIR;
  const resolvedRunDir = runDir || Deno.env.get("ATPROTO_E2E_RUN_DIR") ||
    join(baseDir, resolvedRunId);
  const resolvedLogsDir = Deno.env.get("ATPROTO_E2E_LOG_DIR") || join(resolvedRunDir, "logs");
  const resolvedReportsDir = Deno.env.get("ATPROTO_E2E_REPORTS_DIR") ||
    join(resolvedRunDir, "reports");
  const resolvedDiagDir = diagnosticsDir || Deno.env.get("ATPROTO_E2E_DIAGNOSTICS_DIR") ||
    join(resolvedRunDir, "diagnostics");
  const resolvedPidFile = Deno.env.get("ATPROTO_E2E_PID_FILE") || join(resolvedRunDir, "pids.txt");

  const composeRunId = resolvedRunId.replace(/[^a-z0-9-]+/g, "-");
  const composeProject = Deno.env.get("ATPROTO_E2E_COMPOSE_PROJECT") ||
    `garazyk-e2e-${composeRunId}`;

  for (const path of [resolvedRunDir, resolvedLogsDir, resolvedReportsDir, resolvedDiagDir]) {
    await Deno.mkdir(path, { recursive: true });
  }

  Deno.env.set("ATPROTO_E2E_RUN_ID", resolvedRunId);
  Deno.env.set("ATPROTO_E2E_RUN_DIR", resolvedRunDir);
  Deno.env.set("ATPROTO_E2E_LOG_DIR", resolvedLogsDir);
  Deno.env.set("ATPROTO_E2E_REPORTS_DIR", resolvedReportsDir);
  Deno.env.set("ATPROTO_E2E_DIAGNOSTICS_DIR", resolvedDiagDir);
  Deno.env.set("ATPROTO_E2E_PID_FILE", resolvedPidFile);
  Deno.env.set("ATPROTO_E2E_COMPOSE_PROJECT", composeProject);

  return {
    runId: resolvedRunId,
    runDir: resolvedRunDir,
    logsDir: resolvedLogsDir,
    reportsDir: resolvedReportsDir,
    diagnosticsDir: resolvedDiagDir,
    pidFile: resolvedPidFile,
    composeProject,
  };
}

function redact(value: string): string {
  let redacted = value;
  for (const pattern of SECRET_PATTERNS) {
    redacted = redacted.replace(pattern, (match, p1) => {
      if (match.toLowerCase().includes("accessjwt")) {
        return `${p1}[REDACTED]"`;
      }
      return `${p1}[REDACTED]`;
    });
  }
  return redacted;
}

async function writeText(path: string, text: string) {
  await Deno.mkdir(join(path, ".."), { recursive: true });
  await Deno.writeTextFile(path, redact(text));
}

async function collectHttpEndpoint(
  outputDir: string,
  name: string,
  url: string,
  headers: Record<string, string> = {},
  timeout = 8000,
) {
  const target = join(outputDir, "http", `${name}.txt`);
  let text: string;
  try {
    const controller = new AbortController();
    const id = setTimeout(() => controller.abort(), timeout);
    const resp = await fetch(url, { headers, signal: controller.signal });
    clearTimeout(id);
    const body = (await resp.text()).substring(0, 50000);
    text = `url=${url}\nhttp_status=${resp.status}\ncontent_type=${
      resp.headers.get("Content-Type") || ""
    }\n\n${body}\n`;
  } catch (exc) {
    text = `url=${url}\nerror=${exc}\n`;
  }
  await writeText(target, text);
}

export async function collectDiagnostics(
  context: E2ERunContext,
  options: {
    serviceUrls?: Record<string, string>;
    appviewAdminSecret?: string;
    label?: string;
  } = {},
): Promise<string> {
  const outputDir = context.diagnosticsDir;
  await Deno.mkdir(outputDir, { recursive: true });
  const urls = { ...SERVICE_URLS, ...options.serviceUrls };
  const appviewSecret = options.appviewAdminSecret || Deno.env.get("APPVIEW_ADMIN_SECRET") ||
    "localdevadmin";
  const label = options.label || "atproto-e2e";

  const metadata: Record<string, any> = {
    label,
    run_id: context.runId,
    run_dir: context.runDir,
    diagnostics_dir: outputDir,
    compose_project: context.composeProject,
    created_at_utc: new Date().toISOString(),
    service_urls: urls,
  };

  try {
    const gitHead = new Deno.Command("git", { args: ["rev-parse", "HEAD"] }).outputSync();
    metadata.git_commit = new TextDecoder().decode(gitHead.stdout).trim();
    const gitStatus = new Deno.Command("git", { args: ["status", "--short"] }).outputSync();
    metadata.git_status = new TextDecoder().decode(gitStatus.stdout).trim().split("\n");
  } catch (e) {
    metadata.git_error = String(e);
  }

  await writeText(join(outputDir, "run-metadata.json"), JSON.stringify(metadata, null, 2));

  if (await exists(context.pidFile)) {
    await copy(context.pidFile, join(outputDir, "pids.txt"), { overwrite: true });
  }

  if (await exists(context.logsDir)) {
    const logsOut = join(outputDir, "service-logs");
    await Deno.mkdir(logsOut, { recursive: true });
    for await (const entry of Deno.readDir(context.logsDir)) {
      if (entry.isFile && entry.name.endsWith(".log")) {
        await copy(join(context.logsDir, entry.name), join(logsOut, entry.name), {
          overwrite: true,
        });
      }
    }
  }

  const authHeader = { "Authorization": `Bearer ${appviewSecret}` };

  await Promise.all([
    collectHttpEndpoint(outputDir, "plc-health", `${urls.plc}/_health`),
    collectHttpEndpoint(
      outputDir,
      "pds-describe-server",
      `${urls.pds}/xrpc/com.atproto.server.describeServer`,
    ),
    collectHttpEndpoint(outputDir, "relay-health", `${urls.relay}/api/relay/health`),
    collectHttpEndpoint(outputDir, "relay-upstreams", `${urls.relay}/api/relay/upstreams`),
    collectHttpEndpoint(
      outputDir,
      "appview-backfill-status",
      `${urls.appview}/admin/backfill/status`,
      authHeader,
    ),
    collectHttpEndpoint(
      outputDir,
      "appview-backfill-queue",
      `${urls.appview}/admin/backfill/queue?limit=10`,
      authHeader,
    ),
    collectHttpEndpoint(
      outputDir,
      "appview-ingest-health",
      `${urls.appview}/admin/ingest/health`,
      authHeader,
    ),
    collectHttpEndpoint(
      outputDir,
      "appview-metrics-stats",
      `${urls.appview}/admin/appview/metrics/stats`,
      authHeader,
    ),
    collectHttpEndpoint(
      outputDir,
      "appview-lexicons",
      `${urls.appview}/admin/lexicons`,
      authHeader,
    ),
    collectHttpEndpoint(outputDir, "appview-hooks", `${urls.appview}/admin/hooks`, authHeader),
    collectHttpEndpoint(
      outputDir,
      "appview-endpoints",
      `${urls.appview}/admin/endpoints`,
      authHeader,
    ),
    collectHttpEndpoint(
      outputDir,
      "pds2-describe-server",
      `${urls.pds2}/xrpc/com.atproto.server.describeServer`,
    ),
    collectHttpEndpoint(outputDir, "chat-health", `${urls.chat}/_health`),
    collectHttpEndpoint(outputDir, "video-health", `${urls.video}/_health`),
    collectHttpEndpoint(outputDir, "ui-admin", `${urls.ui}/admin`),
  ]);

  return outputDir;
}
